### ISSUE TITLE
UB for passing uninitialized variables to "significant only at root" arguments

### ISSUE THREAD

**Issue by jprotze** [mpi-forum/mpi-issues#854]
# Problem

The wording about "significant only at root" is a bit sloppy and at least one example is broken. Although the parameters will not be read by the function on non-root ranks, passing uninitialized variables to the function is UB according to the C standard. Example 6.3 does not initialize `rbuf` on non-root ranks and passes this variable to `MPI_Gather`, which is UB. As a result, we observed the `if (myrank == root)` to be optimized away by clang 14 or newer and icx 2022.1 or newer.

# Proposal

- Fix example 6.3 and initialize the variable on declaration.
- Expand the description for _significant only at root_

# Changes to the Text

Change 

```C
int root, myrank, *rbuf;
```
to
```C
int root, myrank, *rbuf=NULL;
```
And possibly explain why this initialization is important?

In the introduction of Chapter 6:

> Some arguments in the collective functions are specified as “significant only at root,” and are ignored for all participants except the root. 

Add:
> **The C standard nevertheless requires variables passed to these arguments to be initialized.**

# Impact on Implementations

None

# Impact on Users

Avoid running into non-obvious UB by misunderstanding “significant only at root”

# References and Pull Requests

**Comment by devreal** [comment]
Curious whether the compiler removed the branch because `rbuf` is uninitialized of because `root` is uninitialized in this example? Agreed that both are UB and we should be careful not to promote that.

**Comment by jprotze** [comment]
In our concrete code, we had root fixed to 0 and more code executed next to the malloc. A `printf` would even print rank as 0 for all ranks:

```C
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
  if (argc != 2) {
    printf("No valid size given\n");
    return 1;
  }

  int n = atoi(argv[1]);

  MPI_Init(&argc, &argv);
  int rank, wsize;
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &wsize);
  n = n - n % wsize;
  int n_per_process = n / wsize;

  float *x;
  float *x_root;

  if (rank == 0) {
    printf("From Rank %d: Initializing vectors...\n", rank);
    fflush(stdout);
    x_root = (float *)malloc(n * sizeof(float));
    memset(x_root, 0, n * sizeof(float));
  }

  x = (float *)malloc(n_per_process * sizeof(float));
  memset(x, 0, n_per_process * sizeof(float));
  MPI_Scatter(x_root, n_per_process, MPI_FLOAT, x, n_per_process, MPI_FLOAT, 0,
              MPI_COMM_WORLD);

  MPI_Finalize();
  free(x);
  return 0;
}
```

**Comment by jeffhammond** [comment]
I don't believe any of the examples are intended to be complete working code. Do they even include all the necessary headers to compile?

**Comment by wgropp** [comment]
That is correct, most examples are code snippets. The code check uses LaTeX comments to describe things like headers and declarations needed to see if the code could be compiled - this test has caught many errors, such as misspellings, out of order arguments, and missing semicolons, in the code snippets. 

However, we do want the code snippets to be valid and compileable with the additional info, so things like ensure rbuf is initialized is reasonable for the examples.

**Comment by mahermanns** [comment]
### Discussion in Tools WG on Sep 30, 2024:

- We should add the suggested text
  - Potentially with "on all participating processes"
- We should also change the "are ignored" to "are ignored by MPI" (or something alike)
  - The MPI implementation is the only thing that we can mandate for
  - What about tools? Do we need additional text to make sure tools don't do "the wrong thing" :tm: 
    - A potential memory checker might want to evaluate the "non-significant" argument anyway

**Comment by mahermanns** [comment]
### Discussion on Tools WG on Oct 14, 2024:

- Chapter (see "MPI_Scatter") the text actually uses "ignored", argument descriptions (in LIS) are using "significant only at root".
    - We should change the "ignored" text to "significant only at" also in the text
    - Define "significant" some place in the text
        - "MPI does not change behavior, based on the value of that argument" ?
            - We should focus on the "behavior" part of the MPI library and not trying to define terms like "ignore" in the context MPI
            - Could be in the realm of both AtoU and AtoI
                - "Users, always initialize your variables!" 
        - Unsure where this should be done? Terms? Collectives?
    - Should bring this up to the Forum (non-voting meeting) to ask how to move forward here
    - Affected chapters:
        - Collectives
        - RMA (MPI_Accumulate)
        - Tools (MPI_T)
        - Process Initialization (MPI_Comm_spawn)

**Comment by jprotze** [comment]
Here is a semantically equivalent reproducer without MPI to look at the compiler optimization observed above on [godbolt](https://godbolt.org/z/MjbsMvKvM). For clang the optimization seems to be introduced with version 14:

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int test(int n, char *c) {
  if (n == 0)
    return 0;
  return atoi(c);
}

int main(int argc, char **argv) {
  char *buf;
  if (argc > 1) {
    buf = strdup(argv[1]);
  }
  int res = test(argc - 1, buf);
  return res;
}
```

**Comment by mahermanns** [comment]
### Tools Meeting 2024-11-22:

- Passing of `buf` to `test(...)` by value is a read on an indeterminate variable, which causes undefined behavior
  - Initialization of `char *buf == NULL` will fix this behavior.
- Prepare straw polls:
  - Should we mentions this at all: yes / no / no opinion
  - Should we provide an AtU: yes / no / no opinion
  - AtU in Terms, and reference at specific places in Collectives / Tools / RMA / ...: yes / no / "verbatim copy" in each location / no opinion (our preferred option would be _yes_)
  - Revise wording for "significant only at"?

**Comment by bkmgit** [comment]
May want to see https://github.com/mpi-forum/mpi-issues/issues/693 - this does not directly address compiler behavior

**Comment by jprotze** [comment]
I'm not sure that a specific version reference matters, since the relevant text didn't change in the last 10+ years. 

C11,C17,C23: list of undefined behavior:
 — An lvalue designating an object of automatic storage duration that could have been declared with the register storage class is used in a context that requires the value of the designated object, but the object is uninitialized. (6.3.2.1).

C23, 6.3.2.1:
If the lvalue designates an object of automatic storage duration that could have been declared with the register storage class (never had its address taken), and that object is uninitialized (not declared with an initializer and no assignment to it has been performed prior to use), the behavior is undefined.

At the end calling out that MPI is compatible with C99 or whatever specific version would also be dangerous, because it either means that compatibility with the other versions is unspecified or will still lead to undefined behavior, if an application developer relies on C99 semantics, where I couldn't find above UB, and compiles with a compiler that assumes newer C semantics.

**Comment by jeffhammond** [comment]
No compiler is perfect and compilers are not required to do bad things wtih UB anyways.  More recent standards of C are more restrictive, so we should reference the latest one that makes sense.

There is no C03, but C99.  C89 is no longer published.  We should use the latest version that is widely availabl.  C11 is widely available, C23 is not.  I recall C17 was a minor change anyways.

Unfortunately, it's too late for https://github.com/mpi-forum/mpi-standard/pull/816 to make it this time, but we should get it in soon anyways.

**Comment by mahermanns** [comment]
The results of my straw-poll is in:

![image](https://github.com/user-attachments/assets/d7f49f46-d708-44a0-8e86-908b5fb5cc0b)

With the majority favoring a both a fix of the examples **and** a brief explanation in the Terms chapter, but it being quite close as to have **all of it** as an errata, I would propose the following:

1. Check and fix any uninitialized buffer variables in "significant at root" examples.
2. Provide a minimal AtU in Terms
3. Add references to this AtU in respective places throughout the standard

And then potentially **revisit** this at a later time, to explain the situation more elaborate (if we think this to be beneficial)

**Comment by devreal** [comment]
I would recommend splitting 1 and 2/3 into separate tickets. I'm worried that "Add references to this AtU in respective places throughout the standard" becomes a change too large for a last-minute errata.

**Comment by mahermanns** [comment]
> I would recommend splitting 1 and 2/3 into separate tickets. I'm worried that "Add references to this AtU in respective places throughout the standard" becomes a change too large for a last-minute errata.

I know what you mean, and we need to carefully craft this *minimal reference*, yet about 60% of the straw poll (I know this does not necessarily equate to real votes on the ticket) would be OK with having some form of brief reference. As we don't have any example code fix in this Issue and PR, I would then create a different issue and PR for that (would be **1** in my previous list)

**Comment by mahermanns** [comment]
It's worth mentioning that there are more places in the standard using the "significant only at root", especially for argument of `MPI_Comm_spawn`, yet it was unclear how to reference the AtU there without major text changes. This should be revisited in a later version of the text.

**Comment by wesbland** [comment]
This passed a [errata vote](https://mpi-forum.org/meetings/2025/01/votes).
<table>
<tr>
    <td> Yes
    <td> No
    <td> Abstain
<tr>
    <td> 30
    <td> 0
    <td> 0
</table>
